source/common/conn_pool/conn_pool_base.cc

```diff
@@ -301,18 +303,16 @@ void ConnPoolImplBase::drainConnectionsImpl() {
   }
 }
 
-void ConnPoolImplBase::checkForDrained() {
-  if (drained_callbacks_.empty()) {
-    return;
+void ConnPoolImplBase::checkForIdle() {
+  if (is_draining_) {
+    closeIdleConnectionsForDrainingPool();
   }
 
-  closeIdleConnectionsForDrainingPool();
-
-  if (pending_streams_.empty() && ready_clients_.empty() && busy_clients_.empty() &&
-      connecting_clients_.empty()) {
-    ENVOY_LOG(debug, "invoking drained callbacks");
-    for (const Instance::DrainedCb& cb : drained_callbacks_) {
-      cb();
+  if (has_seen_clients_ && pending_streams_.empty() && ready_clients_.empty() &&
```

Comment:
is has_seen_clients_ valid for the is_draining case?
AFIK if a cluster has never seen traffic and it is told to drain, it should drain.

I wonder if for clusters we're willing to erase we should simply only create them when there's demand, and if the work between creating the cluster and creating a stream would allow removing this boolean?

Reply:
You're right -- I think we need to drain right away. I'll push an update with a test.

We already create connection pools on demand in `ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::{tcpConnPool,connPool}` -- the bool is there to prevent an immediate cleanup on a new pool when the first idle callback is added (we call `checkForIdle` when a callback is added).

We might be able to remove the bool if we don't check for idle in `addIdleCallback` when `DrainPool::No` is passed to the function. That would also solve the issue of always draining when being told to. WDYT?

Reply:
yeah, if it's just there to avoid immediate drain I think we can probably avoid it with careful calls.  I'm also fine leaving it in as a "but we want to be paranoid"  precaution, with appropriate comments :-)